### ISSUE TITLE
Set _NET_WM_WINDOW_TYPE and related fields properly

### ISSUE THREAD

**Issue by midnightexigent** [elkowar/eww#153]
I am just getting started with eww. I tried to apply the eww-bar example. 

The problem I am encountring is that the "reserve" attribute doesn't seem to do anything: the bar is displaying over the windows. 

Am I missing something ? 

PS. I am using a tiling window manager "leftwm"

**Comment by pradyungn** [comment]
This is known to have happened in the past, and apparently there's not much you can do atm. Personally I suggest looking into your WM's padding properties for the specified monitor, that way you can just pad it to the space you'd normally reserve.

**Comment by midnightexigent** [comment]
bummer.. it is possible to do that but the padding makes so the bar stays visible even when in full-screen mode

**Comment by undefinedDarkness** [comment]
Yeah reserve is a bit so - so at functioning
Sometimes the window's focusable attribute plays a part tho so try setting that to true

**Comment by elkowar** [comment]
yea the issue is that X's struts stuff is _really_ fucky, and every WM does something different pretty much..... I really don't get how other bars do it, other than just implementing custom code per WM

**Comment by undefinedDarkness** [comment]
> other than just implementing custom code per WM
Sounds like very pain

**Comment by midnightexigent** [comment]
I've done some digging, and I am pretty sure that I found what's wrong. 

TL; DR: `eww` is not asking for the desired behavior from the WM in an "EWMH-compliant way". For a status bar, the `_NET_WM_WINDOW_TYPE` property should be set to `_NET_WM_WINDOW_TYPE_DOCK` (what polybar uses) or `_NET_WM_WINDOW_TYPE_TOOLBAR` whereas `eww` is always setting it to `_NET_WM_WINDOW_TYPE_NORMAL` 

How I found out (long version):

comparing the EWMH properties of an `eww` window and `polybar`, we have two notable differences: `_NET_WM_STATE` and `_NET_WM_WINDOW_TYPE`

polybar: 
```bash
$ xprop -name polybar-mainbar0_DP-4
_NET_WM_PID(CARDINAL) = 24959
_NET_WM_DESKTOP(CARDINAL) = 0
_NET_WM_STATE(ATOM) = _NET_WM_STATE_STICKY, _NET_WM_STATE_ABOVE
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_DOCK
WM_CLASS(STRING) = "polybar", "Polybar"
WM_NAME(STRING) = "polybar-mainbar0_DP-4"
_NET_WM_STRUT_PARTIAL(CARDINAL) = 0, 0, 32, 0, 0, 0, 0, 0, 8, 1919, 0, 0
_NET_WM_STRUT(CARDINAL) = 0, 0, 32, 0
```
eww bar:
```bash
 $ xprop -name "Eww - bar"
_NET_WM_DESKTOP(CARDINAL) = 0
_NET_WM_STATE(ATOM) =
WM_HINTS(WM_HINTS):
                Client accepts input or input focus: True
                Initial state is Normal State.
                window id # of group leader: 0x3800001
_GTK_THEME_VARIANT(UTF8_STRING) =
_NET_WM_STRUT_PARTIAL(CARDINAL) = 0, 0, 43, 0, 0, 0, 0, 0, 0, 1919, 0, 0
_NET_WM_STRUT(CARDINAL) = 0, 0, 43, 0
XdndAware(ATOM) = BITMAP
_NET_WM_OPAQUE_REGION(CARDINAL) = 0, 0, 940, 1028
_MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x2, 0x0, 0x0, 0x0, 0x0
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_NORMAL
_NET_WM_SYNC_REQUEST_COUNTER(CARDINAL) = 58725778, 58725779
_NET_WM_USER_TIME(CARDINAL) = 10885614
_NET_WM_USER_TIME_WINDOW(WINDOW): window id # 0x3801591
WM_CLIENT_LEADER(WINDOW): window id # 0x3800001
_NET_WM_PID(CARDINAL) = 32515
WM_LOCALE_NAME(STRING) = "en_US.utf8"
WM_CLIENT_MACHINE(STRING) = "desktop-776A"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                program specified location: 0, 0
                program specified minimum size: 1920 by 43
                program specified maximum size: 1920 by 43
                program specified base size: 0 by 0
                window gravity: NorthWest
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING, _NET_WM_SYNC_REQUEST
WM_CLASS(STRING) = "eww-bar", "eww-bar"
WM_ICON_NAME(STRING) = "Eww - bar"
_NET_WM_ICON_NAME(UTF8_STRING) = "Eww - bar"
WM_NAME(STRING) = "Eww - bar"
_NET_WM_NAME(UTF8_STRING) = "Eww - bar"
```

additionally, [conky](https://github.com/brndnmtthws/conky) seems to achieve the [same behavior](https://raw.githubusercontent.com/leftwm/leftwm-community-themes/master/screenshots/tng.jpg). When I inspect its EWMH properties through `xprop -id <0x...>`, one of the properties is `_NET_WM_WINDOW_TYPE = _NET_WM_WINDOW_TYPE_DOCK`

Also, from the [EWMH spec](https://specifications.freedesktop.org/wm-spec/wm-spec-1.3.html#idm45841378143456): "(_NET_WM_WINDOW_TYPE) SHOULD be used by the window manager in determining the decoration, stacking position and other behavior of the window"

The easy solution seems to add something like 
```xml
<window name="example" type="dock | toolbar | dialog | normal "> <! -- defaults to "normal" -->
...
</window>
```

**Comment by undefinedDarkness** [comment]
How do we do this from within gtk tho iirc gtkrs only exposes a few window types

**Comment by midnightexigent** [comment]
[cnx](https://github.com/mjkillough/cnx) uses [xcb](http://rtbo.github.io/rust-xcb/xcb/index.html) and [xcb_util](https://docs.rs/xcb-util/0.3.0/xcb_util/) 

[xcb_util::ewmh](https://docs.rs/xcb-util/0.3.0/xcb_util/ewmh/index.html) seems to be the right module

**Comment by undefinedDarkness** [comment]
Yeah but gtk ...

**Comment by midnightexigent** [comment]
what about it ?

**Comment by elkowar** [comment]
Given that we already have a good bit of X-specific code in the codebase, adding this wouldn't be an issue, so gtk's limits shouldn't really matter here. also, there will be even more of a module separation once #125 is merged (I really godda make that happen finally, I've been slacking off), so not an issue.

When it comes to API, I would want the following:
- when no `reserve` is set, default to window type `normal`
- when `reserve` _is_ set, default to window type `dock`  (or toolbar if that's the better one, idk)
- have a `windowtype` attribute on the `<window>` tag that may be `normal`, `dock`, `dialog`, `toolbar`

Using a separate tag for each window feels weird, and would make these default behaviours impossible.

I'd definitely very much appreciate a PR here!

**Comment by elkowar** [comment]
@midnightexigent are you gonna be looking into doing a PR here? should i assign you? ^^

**Comment by midnightexigent** [comment]
yea I was gonna start looking into a PR next week. feel free to assign me

would you rather have #125 merged before this comes in ?

**Comment by elkowar** [comment]
Actually yea, #125  should probably be merged before that, or otherwise you start branching off of that current PR (it should be pretty close, just needs to go through final testing and resolve the conflicts

**Comment by elkowar** [comment]
@midnightexigent #125 has been merged! Have fun!

**Comment by elkowar** [comment]
@midnightexigent any updates, are you working on this?

**Comment by midnightexigent** [comment]
actually yea, I am gonna open a PR soon to take the discussion there, I've got the whole X11 properties figured out but I am gonna need some additional input for the API

**Comment by elkowar** [comment]
Closed by #176 - some more work will be done here tho
